### ISSUE TITLE
default use of `plot_spikes()` breaks existing code

### ISSUE THREAD

**Issue by djsaunde** [Hananel-Hazan/bindsnet#105]
Since the recent change to `plot_spikes()`, I've been getting errors of the form:

```
Traceback (most recent call last):
  File "random_network_baseline.py", line 91, in <module>
    output='R')
  File "/home/dan/code/bindsnet/bindsnet/pipeline/__init__.py", line 89, in __init__
    self.plot_data()
  File "/home/dan/code/bindsnet/bindsnet/pipeline/__init__.py", line 198, in plot_data
    self.s_ims, self.s_axes = plot_spikes(self.spike_record)
  File "/home/dan/code/bindsnet/bindsnet/analysis/plotting.py", line 91, in plot_spikes
    for layer in network.layers:
AttributeError: 'dict' object has no attribute 'layers'
```

Could we revert back to the default behavior of accepting a dictionary of spikes to plot? I like the option of plotting from a `Network` instance, but I liked the previous default option.

This is totally up for debate, by the way.

**Comment by djsaunde** [comment]
Also, I think that `plot_spikes()` needs a rewrite. It's super long...

**Comment by hqkhan** [comment]
It's meant to accomodate both sides of the coin. As in, you should be able to throw it `dict` of spikes and it should work all the same. I don't currently have the time to look at it therefore if you have time, you could revert `plotting.py` to the commit where it was the older function. I had a commit that removed the old function therefore the one before that will have the older function untouched because I worked by creating `plot_spikes_new` - a copy of `plot_spikes`. 

I agree about the length but a major part of it basically checking if the inputs were valid or not. We could potentially remove a lot of the assertions if we want to assume that the user will provide us with valid inputs altogether. Or, we can have a different function that checks the validity of inputs which was something I wanted to do but couldn't find time to implement.

**Comment by djsaunde** [comment]
I agree that it should work if you give either argument. The problem is, the function by default interprets the first argument as a `Network` object, and the second as a `Dict[str, torch.Tensor]`, unless you explicitly pass `network=network` or `spikes=spikes` (with the correct types). Perhaps we could just have the first argument be `Union[Network, Dict[str, torch.Tensor]]` (`Union` is the logical OR function of [types](https://docs.python.org/3/library/typing.html)), and remove the second argument, and the type of the first argument can be checked in the function.

I'd argue that checking if the inputs are valid or not aren't really a top priority, and I think it's good Python programming, especially if the errors which result by default are informative. Since we mostly build on top of mature Python libraries, the errors that come out of `torch` and `matplotlib` should mostly be enough to help the user debug their programs. See [this SO post](https://stackoverflow.com/questions/46723564/when-should-i-use-assert-in-python) for related details (the screenshot below is from the Python documentation glossary).

![image](https://user-images.githubusercontent.com/1245942/43431700-35bb7038-943d-11e8-8c5d-db65d39a1c3a.png)

**Comment by hqkhan** [comment]
Gotcha. I agree about checking valid inputs. I will make the appropriate changes once I get the time. Sorry about the delay that this will cause. 

Would you like me to completely remove the functionality of being able to pass network in or do you see yourself potentially using it? We can simply remove it if it's causing annoyances.

**Comment by djsaunde** [comment]
Hm, I guess we should just remove it, to reduce complexity. Sorry about the back and forth.

**Comment by djsaunde** [comment]
Fixed by #108
